### PR TITLE
Zenlink integration (testnet-kintsugi)

### PR DESCRIPTION
**Pull request by gregdhill**
Signed-off-by: Gregory Hill <gregorydhill@outlook.com>

See: #712

- [x] [`zenlink-protocol`](https://github.com/zenlinkpro/Zenlink-DEX-Module/tree/master/zenlink-protocol)
- [x] [`zenlink-stable-amm`](https://github.com/zenlinkpro/Zenlink-DEX-Module/tree/master/zenlink-stable-amm)
- [x] [`zenlink-swap-router`](https://github.com/zenlinkpro/Zenlink-DEX-Module/tree/master/zenlink-swap-router)

### CODE REVIEW

File: primitives/src/lib.rs [hunk 1]
```diff
@@ -477,17 +477,29 @@ create_currency_id! {
...
+
 #[derive(Encode, Decode, Eq, Hash, PartialEq, Copy, Clone, RuntimeDebug, PartialOrd, Ord, TypeInfo, MaxEncodedLen)]
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "std", serde(rename_all = "camelCase"))]
 pub enum CurrencyId {
     Token(TokenSymbol),
     ForeignAsset(ForeignAssetId),
     LendToken(LendTokenId),
+    LpToken(LpToken, LpToken),
+    StableLpToken(StablePoolId),
```
Comment (sander2) [inline primitives/src/lib.rs:497]: Why is a stableLpToken identified by a single id whereas normal lptokens consist of a pair? That seems inconsistent to me. I realize that it's probably dictated by the zenlink pallet, I'm just wondering

File: primitives/src/lib.rs [hunk 2]
```diff
@@ -477,17 +477,29 @@ create_currency_id! {
     }
 }
 
+#[derive(Encode, Decode, Eq, Hash, PartialEq, Copy, Clone, RuntimeDebug, PartialOrd, Ord, TypeInfo, MaxEncodedLen)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "std", serde(rename_all = "camelCase"))]
+pub enum LpToken {
+    Token(TokenSymbol),
+    ForeignAsset(ForeignAssetId),
+    StableLpToken(StablePoolId),
```
Comment (sander2) [inline primitives/src/lib.rs:486]: Is there a usecase for this?

File: Cargo.toml [hunk 3]
```diff
@@ -237,3 +237,6 @@ cumulus-relay-chain-inprocess-interface  = { git = "https://github.com/paritytec
 cumulus-relay-chain-rpc-interface  = { git = "https://github.com/paritytech//cumulus", branch = "polkadot-v0.9.31" }
 cumulus-client-collator  = { git = "https://github.com/paritytech//cumulus", branch = "polkadot-v0.9.31" }
 cumulus-relay-chain-minimal-node  = { git = "https://github.com/paritytech//cumulus", branch = "polkadot-v0.9.31" }
+
+[patch."https://github.com/open-web3-stack/open-runtime-module-library"]
+orml-traits = { git = "https://github.com/open-web3-stack//open-runtime-module-library", rev = "3fcd3cf9e63fe80fd9671912833a900ba09d1cc0" }
```
Comment (sander2) [inline Cargo.toml:242]: Fyi we can do better actually now that [workspace inheritance](https://betterprogramming.pub/workspace-inheritance-in-rust-65d0bb8f9424) is a thing. But for now this is fine

**Review by sander2 — APPROVED** [review]
Mostly lgtm - it shouldn't really impact mainnet so if you're confident that the currencyid is final, I'm ok with merging this

File: Cargo.toml (hunk 3 above)
Comment (gregdhill) [reply Cargo.toml:242]: Nice! I didn't realize this was a thing, we should definitely adopt that going forward

File: primitives/src/lib.rs (hunk 1 above)
Comment (gregdhill) [reply primitives/src/lib.rs:497]: Technically the stable pool can support multiple assets whereas the standard AMM can only ever support pairs. However, I see no reason why both couldn't be implemented in the same way but it is probably more effort than is worth it right now for Zenlink to adopt the same interface.

File: primitives/src/lib.rs (hunk 2 above)
Comment (gregdhill) [reply primitives/src/lib.rs:486]: Following the interface adopted by Acala: https://github.com/AcalaNetwork/Acala/blob/1d7852f92a5e1116d8f1ad0e3d0d9e5e2d59c230/primitives/src/currency.rs#L460

**Pull request merged**